opm/simulators/linalg/bda/BILU0.cpp

```diff
@@ -479,117 +495,65 @@ namespace bda
-            int offsetLU = diagIndex[i] + 1;
-            memcpy(Umat->nnzValues + offsetU * bs * bs, LUmat->nnzValues + offsetLU * bs * bs, sizeof(double) * bs * bs * rowSize);
-            memcpy(Umat->colIndices + offsetU, LUmat->colIndices + offsetLU, sizeof(int) * rowSize);
-            offsetU += rowSize;
-            Umat->rowPointers[URowIndex + 1] = offsetU;
-            URowIndex++;
-        }
-#endif
         if (verbosity >= 3) {
             std::ostringstream out;
-            out << "BILU0 decomposition: " << t_decomposition.stop() << " s";
+            out << "BILU0 copy to GPU: " << t_copyToGpu.stop() << " s";
             OpmLog::info(out.str());
         }
 
-        Timer t_copyToGpu;
-        if (pattern_uploaded == false) {
-            queue->enqueueWriteBuffer(s.Lcols, CL_TRUE, 0, Lmat->nnzbs * sizeof(int), Lmat->colIndices);
-            queue->enqueueWriteBuffer(s.Ucols, CL_TRUE, 0, Umat->nnzbs * sizeof(int), Umat->colIndices);
-            queue->enqueueWriteBuffer(s.Lrows, CL_TRUE, 0, (Lmat->Nb + 1) * sizeof(int), Lmat->rowPointers);
-            queue->enqueueWriteBuffer(s.Urows, CL_TRUE, 0, (Umat->Nb + 1) * sizeof(int), Umat->rowPointers);
-            pattern_uploaded = true;
+        Timer t_decomposition;
+        std::ostringstream out;
+        for (int color = 0; color < numColors; ++color) {
+            const unsigned int firstRow = rowsPerColorPrefix[color];
+            const unsigned int lastRow = rowsPerColorPrefix[color+1];
+            const unsigned int work_group_size2 = 128;
+            const unsigned int num_work_groups2 = 1024;
+            const unsigned int total_work_items2 = num_work_groups2 * work_group_size2;
+            const unsigned int num_hwarps_per_group = work_group_size2 / 16;
+            const unsigned int lmem_per_work_group2 = num_hwarps_per_group * bs * bs * sizeof(double);           // each block needs a pivot
+            if (verbosity >= 4) {
+                out << "color " << color << ": " << firstRow << " - " << lastRow << " = " << lastRow - firstRow << "\n";
+            }
+            cl::Event event = (*ilu_decomp_k)(cl::EnqueueArgs(*queue, cl::NDRange(total_work_items2), cl::NDRange(work_group_size2)), firstRow, lastRow, s.LUvals, s.LUcols, s.LUrows, s.invDiagVals, s.diagIndex, LUmat->Nb, cl::Local(lmem_per_work_group2));
+            event.wait();
         }
-        queue->enqueueWriteBuffer(s.Lvals, CL_TRUE, 0, Lmat->nnzbs * sizeof(double) * bs * bs, Lmat->nnzValues);
-        queue->enqueueWriteBuffer(s.Uvals, CL_TRUE, 0, Umat->nnzbs * sizeof(double) * bs * bs, Umat->nnzValues);
-        queue->enqueueWriteBuffer(s.invDiagVals, CL_TRUE, 0, Nb * sizeof(double) * bs * bs, invDiagVals);
+
         if (verbosity >= 3) {
-            std::ostringstream out;
-            out << "BILU0 copy to GPU: " << t_copyToGpu.stop() << " s";
+            out << "BILU0 decomposition: " << t_decomposition.stop() << " s";
             OpmLog::info(out.str());
         }
```

Comment:
This `t_decomposition` measures the kernel duration for exact decomp, `t_copyToGpu` is separate.
ChowPatelIlu does not have any timers right now.
It can be confusing that `chow_patel_decomposition()` is a new function, but the exact decomp is performed in `create_preconditioner()` itself.

Reply:
Should we add a timer for chow-patel, then? Might be interesting to see.

Reply:
87e3f1d